File: lib/ai/services/conversation_manager.dart
import 'package:smartfarm_ai/ai/models/chat_message.dart';

/// Gestiona memoria temporal de conversación (optimiza tokens y mantiene contexto).
class ConversationManager {
  const ConversationManager({this.maxMessages = 12});

  final int maxMessages;

  /// Conserva los últimos N mensajes alternando user/assistant.
  List<ChatMessage> trimForApi(List<ChatMessage>? history) {
    if (history == null || history.isEmpty) return const [];
    if (history.length <= maxMessages) return List.unmodifiable(history);
    return List.unmodifiable(history.sublist(history.length - maxMessages));
  }

  /// Texto resumido del historial para prompts locales.
  String buildContextSummary(List<ChatMessage>? history) {
    final trimmed = trimForApi(history);
    if (trimmed.isEmpty) return '';
    return trimmed
        .map((m) => '${m.role == 'user' ? 'Usuario' : 'Asistente'}: ${m.content}')
        .join('\n');
  }
}

File: backend/src/ai/conversationManager.js
/**
 * Gestión de historial conversacional (memoria temporal, optimización de tokens).
 */
const MAX_MESSAGES = 12;

function trimHistory(history) {
  if (!Array.isArray(history) || history.length === 0) return [];
  const sanitized = history
    .filter((m) => m && (m.role === 'user' || m.role === 'assistant') && m.content)
    .map((m) => ({
      role: m.role,
      content: String(m.content).trim().slice(0, 2000),
    }))
    .filter((m) => m.content.length > 0);

  if (sanitized.length <= MAX_MESSAGES) return sanitized;
  return sanitized.slice(sanitized.length - MAX_MESSAGES);
}

function buildContextSummary(history) {
  const trimmed = trimHistory(history);
  if (trimmed.length === 0) return '';
  return trimmed
    .map((m) => `${m.role === 'user' ? 'Usuario' : 'Asistente'}: ${m.content}`)
    .join('\n');
}

module.exports = { trimHistory, buildContextSummary, MAX_MESSAGES };
